### ISSUE TITLE
[FR] Support multiple Label printers within the same label plugin

### ISSUE THREAD

**Issue by wolflu05** [inventree/InvenTree#4789]
### Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find a similar feature request

### Problem statement

The current label printing system is not very scalable in larger environments with multiple printers of the same brand because each plugin only has the ability to expose one printer.

### Suggested solution

The current system has a huge downside. Every plugin can only provide one printer. I thought very long about how we can fix that system without the downsides listed in alternatives I've already considered. This is my final concept:

## 💡 The idea

1. We add a new `PrintingDriverMixin` which lets the developer define needed settings for **each** (physical) printer. That should work somehow the same way as the current `SettingsMixin`, but is scoped to a specific (physical) printer
2. InvenTree settings get a new Tab `Printers` to manage printers. We should think about a printing plugin more like a provided printing driver than just a function. In that tab we can manage printers. A printer has the following attributes:
   - name
   - driver (select field - plugin that provides that driver)
   - defined config for that specific printer (specified in 1)
   - (scopes)
   - ...
3. We add printers by giving it a name, selecting a driver and fill out the driver specific settings
4. When printing we can choose between all created printers. The printing plugin then gets called with the pdf that needs to be printed and the Printer model which contains the printer specific settings.

This changes the way printing plugins work, so we need some migration period to allow the migrating of the old `LabelPrintingMixin` plugins. 

## 🚀 Future ideas
This enables us much more possibilities in the future:
- printers can be sorted based on priority for a specific label that need to be printed (e.g. that paper box labels are printed with a different printer than labels for magazine boxes)
- printers can be scoped to locations, categories, ... (you're most like not using a printer that is in Fab1 while you are in Fab2)

## 👍 Advantages

- ✅ Easily scaleable for large environments
- ✅ Not every developer has to think about handling multiple printers on themself

### Describe alternatives you've considered

1. Allow plugins to be instantiated multiple times with different settings (Ref: #3568) 
   - ❌ not very durable and can cause many bugs
2. Allow plugins to expose multiple printers
    - ❌ every printer plugin dev needs to handle multiple printers support on their own

### Do you want to develop this?

- [ ] I want to develop this.

**Comment by SchrodingersGat** [comment]
On the face of it, this approach seems a bit "heavy handed". For discussion, I'd like to propose a simpler alternative.

Each `LabelPrintingMixin` plugin exposes a method `get_printers()` which returns a list of *names* of available printers. These can be enumerated in whatever way the individual plugin wants. By default, the `get_printers()` plugin returns a single printer, as the current implementation, so existing printer plugins are not impacted.

So a plugin may decide it wants to expose multiple printers, and can also decide how to name them for the user to select. Then the printers are exposed to the user in the current fashion.

This does not provide advanced functionality like print queues / scopes / advanced generic configuration. But it would be much simpler to implement.

I'm not sure how much effort to put into advanced printing functionality - how widely is this being used?

**Comment by matmair** [comment]
I would like to have proper machine (in this case printers) management and would like to work on this if @wolflu05 is not taking dips. A first implementation could be pretty simple. I am imagining something like the notification method registry. That works pretty great IMO. Something like this but for machines, with plugins providing machine types. The new printing plugin would then implement a base LabelPrintingMachine - whoes core function is the same printing function as always. The internal APIs from other parts of InvenTree need to change very little to support this.

The steps for implementing this would be:
- Add a registry for machines
- Add a model for holding machine configs (just like InvenTreePlugins and PluginConfigs work)
- Add a BaseMachine
- Add a BaseLabelPrintingMachine
- Add management APIs and an API driven -ui for -machines (of all types)

@SchrodingersGat what do you think?

**Comment by matmair** [comment]
Suggested name change by @wolflu05 is 'Driver' - which I find more fitting than Macine. The function remains the same. I edited my suggestion acordingly.

**Comment by wolflu05** [comment]
1. Can you maybe point out the differences to my approach a bit more? 
2. So do I get this right, your machines are a second, different type of „plugins“ that can’t be used together. E.g. use the current settings mixin, app mixin with a machine? 
3. And regarding implementation, I‘d got a bit more comfortable with Django and thought this could be my first try for a bigger feature, but if you say you want to develop that now, go ahead, you know the codebase better than I. I don’t think I can get your approach with that new type of plugin registry to work clean. 
4. Can you maybe also explain more how you imagine your PnP machine to integrate with this new machine registry, maybe this explains your general thoughts more.
5. are individual settings for each machine defined at the machine type also included in your idea? And how do you imagine the userinteface, we click on a button add machine, then Chose machine type (pnp, printer, …) and then select an available driver? (Btw I kind of like that name to call them „driver“) and then enter the driver required settings like  ip address, username, …

**Comment by wolflu05** [comment]
2. I meant from the python implementation side
3. That sounds like a good plan. I'll start and may ask you if I have a specific question
5. I'm not sure if we should call it a driver registry. I actually liked the name machine registry, but I think the thing that creates the bridge between machine API and inventree should be called driver. I'm not sure how good that is in term of naming conventions and consistency.

## new question
6.Where do you want that implementation to live? Distributed via pypi? What are the advantages, that would complicate things more I guess. Shouldn't that be its own Django app in core?

**Comment by matmair** [comment]
2. The machine/driver could probably use some methods like settings as pointed out in 5. I do not think every mixin makes sense
3. The backend is probably a lot simpler than the UI, I imagine it being similar to MethodStorageClass and NotificationMethod. The only thing I know about figma is that they belong to Adobe - If it can help why not
4. We can also call it a machine registry. I am not too hawkish on any name rn. The nomenclature should probably be `Machine` => 1 instance (in database) of a thing that needs a `Driver` that facilitates communication between the Inventree `Machine` and real-world `Machine` ie thing

6. There was no positive feedback from the maintainer. I think we should wait at least a few days for feedback on the points brought up. I would like to see it in core. It could live in a plugin too. There is no real downside to that (except maybe for some install hurtles)

**Comment by SchrodingersGat** [comment]
As discussed over on discord, @wolflu05 if you want to take the lead on this I'd be happy to review and assist :)

**Comment by wolflu05** [comment]
I'd made a first [draft](https://design.penpot.app/#/view/248140e0-d822-80a6-8002-7b2c566cd1db?page-id=248140e0-d822-80a6-8002-7b2c566cd1dc&section=interactions&index=0&share-id=5d8c4d0a-1e64-8116-8002-7b2e09ad7b91) with penpot of how the userinterface could look like. What approach do you think would be the best (V1,V2 or V3)?

I also thought a bit more about the plugin system implementation and think that the current mixins doesn't make sense to be used with a machine, which would enable use to only discover drivers and there is no need to activate them a plugin inoder to see them appearing in the machine driver selection. So the `InvenTreePlugin` class is not needed for a machine. I image the API like the following. What do you think?

```py
from zebra import ZebraPrinter # dummy zebra library

class ZebraPrinterDriver(PrintingDriverMixin, BaseDriver):
    """Zebra label printing machine driver"""

    NAME = "Zebra printing driver"
    VERSION = "1.0"
    AUTHOR = "wolflu05"

    def print_page(self, machine, pdf_data):
        """Send the label to the printer"""
        zebra_printer = ZebraPrinter(ip=machine.settings.ip, user=machine.settings.ip)
        zebra_printer.print(pdf_data)
```

**Comment by matmair** [comment]
@wolflu05 nice tool. I think V1 is the cleanes / easiest to work with.
As for the python code: I think the notification methods are pretty much how it could work. So a plugin would hold 1+ Drivers, discovered via inheritance. Maybe take a look at [this](https://github.com/matmair/inventree-apprise/blob/main/src/inventree_apprise/ApprisePlugin.py). I think Drivers could work this way. That would enable global settings for a driver (via plugin settings) and per-machine settings.

**Comment by wolflu05** [comment]
This is much better. Thank you I'll take a look at that part - haven't looked into notifications yet.

**Comment by wolflu05** [comment]
Can you maybe give me a brief overview on where you would put what part? Should machines be their own app, should they live in the plugin app or common app?

**Comment by matmair** [comment]
```
.
- InvenTree
- - machine
- - - drivers: Here are all building drivers
- - - - __init__.py: Clean import for all base drivers
- - - - BaseLabelPrintingDriver.py
- - - - test_BaseLabelPrintingDriver.py
- - - driver.py: Here is the base stuff for drivers: The base driver and helpers (if needed) [sample](https://github.com/inventree/InvenTree/blob/306f36bff8f3326be0b32eb0f083858a259329c4/InvenTree/common/notifications.py#L21-L146)
- - - registry.py: Here is the registration logic as a class and 1  instance of the class as permanent storage - [sample](https://github.com/inventree/InvenTree/blob/306f36bff8f3326be0b32eb0f083858a259329c4/InvenTree/common/notifications.py#L166-L235)
- - - apps.py: Things that should run at the start of django ([AppConfig](https://github.com/inventree/InvenTree/blob/306f36bff8f3326be0b32eb0f083858a259329c4/InvenTree/InvenTree/apps.py#L21-L47) to [collect drivers](https://github.com/inventree/InvenTree/blob/306f36bff8f3326be0b32eb0f083858a259329c4/InvenTree/InvenTree/apps.py#L197-L201) and init all machines from the model)
- - - api.py: API routes and functions
- - - admin.py: Small admin interface for mass management (activate/deactivate/delete/ ... room for more)
- - - models.py: Model to hold configs for machines/drivers
- - - urls.py
- - - views.py: Frontend functions if any (would this work with preact - ie just a template and no custom views?)
```
@wolflu05 something like this should work. I have not spelled out everything so if something is unclear please ask anytime. Also open to feedback - this is set up to work easily with a much larger bank of BaseDrivers etc.

**Comment by wolflu05** [comment]
Thanks a lot, this gives me a basic structure to work with.

**Comment by wolflu05** [comment]
Still valid, See #4824

**Comment by wolflu05** [comment]
In development at #4824, currently waiting on PUI and a few missing methods in the form framework.

**Comment by wolflu05** [comment]
Currently in development and making some progress with the Machine admin in PUI. (#4824)

**Comment by wolflu05** [comment]
Still wip in #4824, I know self assigned me this.
